pgjdbc/src/main/java/org/postgresql/ssl/PKCS12KeyManager.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2019, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.ssl;
+
+import org.postgresql.util.GT;
+import org.postgresql.util.PSQLException;
+import org.postgresql.util.PSQLState;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.net.Socket;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.Principal;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+
+import javax.net.ssl.X509KeyManager;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.x500.X500Principal;
+
+public class PKCS12KeyManager implements X509KeyManager {
+
+  private final CallbackHandler cbh;
+  private PSQLException error = null;
+  private final String keyfile;
+  private final KeyStore keyStore;
+  boolean keystoreLoaded = false;
+
+  public PKCS12KeyManager(String pkcsFile, CallbackHandler cbh) throws PSQLException {
+    try {
+      keyStore = KeyStore.getInstance("pkcs12");
+      keyfile = pkcsFile;
+      this.cbh = cbh;
+    } catch ( KeyStoreException kse ) {
+      throw new PSQLException(GT.tr(
+        "Unable to find pkcs12 keystore."),
+        PSQLState.CONNECTION_FAILURE, kse);
+    }
+  }
+
+  /**
+   * getCertificateChain and getPrivateKey cannot throw exeptions, therefore any exception is stored
+   * in {@link #error} and can be raised by this method.
+   *
+   * @throws PSQLException if any exception is stored in {@link #error} and can be raised
+   */
+  public void throwKeyManagerException() throws PSQLException {
+    if (error != null) {
+      throw error;
+    }
+  }
+
+  @Override
+  public String[] getClientAliases(String keyType, Principal[] principals) {
+    String alias = chooseClientAlias(new String[]{keyType}, principals, (Socket) null);
+    return (alias == null ? new String[]{} : new String[]{alias});
+  }
+
+  @Override
+  public String chooseClientAlias(String[] strings, Principal[] principals, Socket socket) {
+    if (principals == null || principals.length == 0) {
+      // Postgres 8.4 and earlier do not send the list of accepted certificate authorities
+      // to the client. See BUG #5468. We only hope, that our certificate will be accepted.
+      return "user";
+    } else {
+      // Sending a wrong certificate makes the connection rejected, even, if clientcert=0 in
+      // pg_hba.conf.
+      // therefore we only send our certificate, if the issuer is listed in issuers
+      X509Certificate[] certchain = getCertificateChain("user");
```

Comment:
Is there a good reason to hard-code this literal?
I think it must be moved to a static final constant with explanation then.

Reply:
no good reason